### PR TITLE
chore: new models and migrations

### PR DESCRIPTION
**Pull request by Carl-Tabuso**
### Newly added models and migrations for:
- Creation of exams.
- Assigning examination results for each job application.
- Assigning educational and experience requirements for each job title.
- Tracking application's initial and final interview -- including the rating of different parameters for final interview.
- Consolidating documents for each application. Instead of linking these pre-employment reqs/docs to the applicants, it should link to the application table for a central point of access.
- Assigning hard and soft skills for different job titles. It is a pivot table --- other jobs may share the same skills and some skills might be applicable to other jobs, that's why.

### Modified models and migrations for:
- Proper arrangement of table columns. The two-factor secrets should be placed before `created_at` and `updated_at`
- Additional information gathering from the applicants, such as contact, education, etc.
- Normalization of tables.

## Docs
- Add comments before each method to further understand their use-case.

### CODE REVIEW

File: database/migrations/2024_09_03_113160_create_applications_table.php [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                ->cascadeOnDelete();
+
+            $table->foreignIdFor(Application::class, 'application_id')
+                ->constrained('applications', 'application_id')
+                ->cascadeOnUpdate()
+                ->cascadeOnDelete();
+
+            $table->string('file_path');
+            $table->boolean('is_submitted')->default(false);
+            $table->timestamp('submitted_at')->nullable();
```
Comment (C-Ivan-Bandilla) [inline database/migrations/2024_09_03_113160_create_applications_table.php:66]: What if to determine if document is submitted is if  submitted_at is null or has time? just asking

Comment (C-Ivan-Bandilla) [reply database/migrations/2024_09_03_113160_create_applications_table.php:66]: ganun kc sa user verified_at nalalmang verified pag inde null

Comment (C-Ivan-Bandilla) [reply database/migrations/2024_09_03_113160_create_applications_table.php:66]: Whats difference between this $table->timestamps(); and submitted_at

**Review by C-Ivan-Bandilla — APPROVED** [review]

**Comment by Carl-Tabuso** [comment]
To determine the timestamp the submission was made, it will be handled using controller/livewire endpoints. It was initially set to null since we might have to anticipate that an applicant has yet to upload each pre-employment requirement. We can manually set the assignment of values for the `submitted_at` attribute.

Use-case: Applicant uploads a document and upon clicking the submit button, an endpoint will handle the validation, hashing, and insertion of file in the `file_path` attribute. If all of which is successful, assign the timestamp of successful submission in `submitted_at`.

You raise a valid point about the purpose of `$table->timestamps().` I will just replace it with an `updated_at` attribute instead -- if there are changes to the documents.

**Comment by Carl-Tabuso** [comment]
I kind of overlooked the `is_submitted` attribute. You're right that we can just determine if a document is submitted if `submitted_at` isn't null. With this, together with the `$table->timestamps` redundancies, I propose this change:

Schema changes in `application_docs`:

image

Running a fresh migration:

image

**Comment by Carl-Tabuso** [comment]
I might also have to add another attribute for preemp status. I went to the mockup UI and realized that the hr can request a resubmission of pre-employment reqs/docs. Imma do this later on. Anyway before I merge, @C-Ivan-Bandilla do tell me on this thread if the change I proposed addresses your questions.

**Comment by C-Ivan-Bandilla** [comment]
Just a notice on models if there are custom names for timestamp model should be instructed for custom names
just like custom name for primary keys


I think I had used this reference in old versions of the migrations

https://stackoverflow.com/questions/25010435/laravel-custom-timestamp-column-names

**Comment by Carl-Tabuso** [comment]
The implementation to tell eloquent of custom timestamp columns remains the same as the answer given in the StackOverflow thread. 

Laravel docs:

image

I'll do this in a bit and check if this is also the case with other tables.

**Comment by Carl-Tabuso** [comment]
It's nice that you've noticed that such implementation is possible. This reply then is rather obsolete since eloquent itself can determine when the submission and modification were made -- we don't need to manually assign this on our controller endpoints.

> To determine the timestamp the submission was made, it will be handled using controller/livewire endpoints. It was initially set to null since we might have to anticipate that an applicant has yet to upload each pre-employment requirement. We can manually set the assignment of values for the submitted_at attribute.

> Use-case: Applicant uploads a document and upon clicking the submit button, an endpoint will handle the validation, hashing, and insertion of file in the file_path attribute. If all of which is successful, assign the timestamp of successful submission in submitted_at.

**Comment by Carl-Tabuso** [comment]
while I was at it, I realized that some of the tables were not normalized. I bunch all the processes in a single table in some migration files. Consider this:

```php
        Schema::create('employee_leaves', function (Blueprint $table) {
            $table->id('emp_leave_id');

            $table->foreignIdFor(Employee::class, 'employee_id')
                ->constrained('employees', 'employee_id')
                ->cascadeOnUpdate()
                ->cascadeOnDelete();

            $table->foreignIdFor(LeaveCategory::class, 'leave_id')
                ->constrained('leave_categories', 'leave_id')
                ->cascadeOnUpdate()
                ->cascadeOnDelete();

            $table->longText('reason');
            $table->timestamp('start_date');
            $table->timestamp('end_date');
            $table->boolean('is_supervisor_approved')->default(false);
            $table->timestamp('supervisor_approved_at')->nullable();

            $table->foreignIdFor(Employee::class, 'supervisor')
                ->constrained('employees', 'employee_id')
                ->cascadeOnUpdate()
                ->cascadeOnDelete();

            $table->boolean('is_area_man_approved')->default(false);
            $table->timestamp('area_man_approved_at')->nullable();

            $table->foreignIdFor(Employee::class, 'area_manager')
                ->constrained('employees', 'employee_id')
                ->cascadeOnUpdate()
                ->cascadeOnDelete();

            $table->boolean('is_hr_manager_approved')->default(false);
            $table->timestamp('hr_manager_approved_at')->nullable();

            $table->foreignIdFor(Employee::class, 'hr_manager')
                ->constrained('employees', 'employee_id')
                ->cascadeOnUpdate()
                ->cascadeOnDelete();

            $table->timestamps();
        });

```

As you may noticed, this table stores every procedure from supervisor approval up to hr. This is also the case with performance evaluations and overtimes tables, they have these exact same columns:

```php
            $table->boolean('is_supervisor_approved')->default(false);
            $table->timestamp('supervisor_approved_at')->nullable();

            $table->foreignIdFor(Employee::class, 'supervisor')
                ->constrained('employees', 'employee_id')
                ->cascadeOnUpdate()
                ->cascadeOnDelete();

            $table->boolean('is_area_man_approved')->default(false);
            $table->timestamp('area_man_approved_at')->nullable();

            $table->foreignIdFor(Employee::class, 'area_manager')
                ->constrained('employees', 'employee_id')
                ->cascadeOnUpdate()
                ->cascadeOnDelete();

            $table->boolean('is_hr_manager_approved')->default(false);
            $table->timestamp('hr_manager_approved_at')->nullable();

            $table->foreignIdFor(Employee::class, 'hr_manager')
                ->constrained('employees', 'employee_id')
                ->cascadeOnUpdate()
                ->cascadeOnDelete();
```


Now, I'm thinking of creating a "process" table that has a polymorphic relationship with other tables. Here, we can store the process_type and process_id. Hypothetically speaking, if an employee made an overtime request, we can insert a new row with a value of "employee_overtime_request" for the process_type and the "employee_overtime_id" for the process _id.

**Comment by C-Ivan-Bandilla** [comment]
I think its okay and I think it can ease the assigning permission of view request and approval

**Comment by Carl-Tabuso** [comment]
Yes, exactly! This is what the `processes` table looks like:

```php
Schema::create('processes', function (Blueprint $table) {
    $table->id('process_id');
    $table->morphs('processable');

    $table->boolean('is_supervisor_approved')->default(false);
    $table->timestamp('supervisor_approved_at')->nullable();

    $table->foreignIdFor(Employee::class, 'supervisor')
        ->nullable()
        ->constrained('employees', 'employee_id')
        ->cascadeOnUpdate()
        ->cascadeOnDelete();

    $table->boolean('is_area_manager_approved')->default(false);
    $table->timestamp('area_manager_approved_at')->nullable();

    $table->foreignIdFor(Employee::class, 'area_manager')
        ->nullable()
        ->constrained('employees', 'employee_id')
        ->cascadeOnUpdate()
        ->cascadeOnDelete();

    $table->boolean('is_hr_manager_approved')->default(false);
    $table->timestamp('hr_manager_approved_at')->nullable();

    $table->foreignIdFor(Employee::class, 'hr_manager')
        ->nullable()
        ->constrained('employees', 'employee_id')
        ->cascadeOnUpdate()
        ->cascadeOnDelete();
});
```

I registered the models as well to be represented as strings in the `AppServiceProvider`

```php
    public function boot(): void
    {
        // other bootstrap services

        Relation::enforceMorphMap([
            'performance_evaluation' => 'App\Models\PerformanceEvaluation',
            'overtime' => 'App\Models\Overtime',
            'employee_leave' => 'App\Models\EmployeeLeave',
        ]);
    }
```

**Comment by Carl-Tabuso** [comment]
It also helps removing bunch of codes about such processes on the `Employee` model.


https://github.com/user-attachments/assets/6e872d23-dccd-4e77-8be0-e840d70b6fb9

**Comment by Carl-Tabuso** [comment]
I think these changes are sufficient enough. We can work on any improvements along the way. Your say @C-Ivan-Bandilla ?

**Comment by C-Ivan-Bandilla** [comment]
I s this area_manager nullable?

image

 after this I think it can be merge

**Comment by Carl-Tabuso** [comment]
tbh, i don't know. From the POSITION.pdf provided by the client, the specific areas are their different branches. Typically, there should always be an area manager.

**Comment by C-Ivan-Bandilla** [comment]
Okay I will merge it, in order to also change my default merge option

**Pull request merged**